### ISSUE TITLE
r.login(username, password) not recognizing that username and password variables have been passed to the function

### ISSUE THREAD

**Issue by georgefahmy** [jmfernandes/robin_stocks#348]
I have a script that has my username and password variables stored in environment variables. at the start of the script, i load those variables, and store them in USERNAME, PASSWORD respectively. 

I am using crontab to execute shell scripts that call my python scripts periodically (several times a day). Some days the login will fail, when the authentication expires. However, the issue is this part of the function: 

`# Try to log in normally.
    if not username:
        username = input("Robinhood username: ")
        payload['username'] = username

    if not password:
        password = getpass.getpass("Robinhood password: ")
        payload['password'] = password`

username and password are valid because i pass them to the function at the beginning. So even though the authentication is not valid, which results in this part of the code executing: 
`except:
                print(
                    "ERROR: There was an issue loading pickle file. Authentication may be expired - logging in normally.", file=get_output())
                set_login_state(False)
                update_session('Authorization', None)`

the username and password checks should not fail, and thus the `input` and `getpass` functions should not be called at all. 

Currently this logic messes up my automation 😞. This issue seems only to happen when the script is executed via cron, but not when I execute the same shell script manually (`./shell_script/run.sh`).

Has anyone seen anything like this?

**Comment by georgefahmy** [comment]
> set SHELL=(whatever shell you want) in your crontab file, like this one: https://unix.stackexchange.com/a/89596.

Thanks for the response. I was just double checking my login function (modified from the built in .login function) and it does indeed call for environment variables. I will give this a try and share an update. 

I am curious as to why it only happens sometimes. and once it happens in one of the scripts, it happens in all of them until i manually re-login (by running the script myself) then it works.

**Comment by babbldev** [comment]
Yeah, let me know how it goes! We run a lot of stuff on cron jobs, and the best way I've found to keep it happy is to treat it like it's running on a completely different computer, so you have to spoon-feed it files, data, and variables.

**Comment by georgefahmy** [comment]
update for you @babbldev. I tried added the `source $HOME/.bashrc` in the shell script and added the `SHELL=/bin/bash` to the crontab but that didnt seem to work. what did work was exporting the variables directly in the shell script that ran the python script. So either the source $HOME/.bashrc isnt work, or the virtual environment activation is messing up that source somehow. I'll need to figure out a solution to this, as i don't want to house the login information in the shell script directly.

**Comment by babbldev** [comment]
Okay perfect, I think you're there. If you only need to access env variables in your .py script, then your best bet is to export them as part of your virtualenv (see section 'Store Environment Variables in Virtual Environment' here: https://medium.com/@dataproducts/python-three-different-ways-to-store-environment-variables-15224952f31b). If you need to access them outside of the python script, you have to make sure you have the `export ENV_VAR=xxxxx` calls in your bash profile (I *think* these env vars are accessible from the venv but that plus cron makes for a confusing scope lol). 

But ideally, only python needs the env vars and you're golden, just plug 'em into the virtualenv activate script.
